hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapReplicationStateHolder.java

```diff
@@ -109,6 +111,14 @@ void applyState() {
                 recordStore.reset();
                 recordStore.setPreMigrationLoadedStatus(loaded.get(mapName));
 
+                MapContainer mapContainer = recordStore.getMapContainer();
```

Comment:
It's the replication. The addIndex operation is executed on the main replica only. With global index, the indexes are in the MapContainer anyway, so each replica can access this knowledge. With index-per-partition, the replica has no way of knowing which indexes have been added in runtime.